File: Python/PEC.py
import os
from numpy import arange
import torch
import torchvision.transforms as transforms
from Function import *

# Underexposure correction parameter example
path = r'./Input/under'
outpath = r'./Result'
colorspace = 'HSV'  # 'RGB' or 'HSV' 
indicator = 0.55 # c
outer_itern = 4   # T
inner_itern = [1, 1, 1, 1]  # K
isinner = 1

'''
# Overexposure correction parameter example
path = r'./Input/over'
outpath = r'./Result'
colorspace = 'RGB'  # 'RGB' or 'HSV' 
indicator = 0.65 # c
outer_itern = 2   # T
inner_itern = [2, 2]  # K
isinner = 1
'''


def pec(img_path, save_path, prompt):
    img = Image.open(img_path)
    img = img.convert('RGB')
    if colorspace == 'RGB' :
        imgx = transforms.ToTensor()(img).unsqueeze(0).cuda()
        y = imgx
    elif colorspace == 'HSV' :
        is_transition = True
        im = np.array(img)
        img = cv2.cvtColor(im, cv2.COLOR_RGB2HSV)
        imgx = transforms.ToTensor()(img).unsqueeze(0).cuda()
        y = imgx[:, 2, :, :].unsqueeze(0)
        img_HS = imgx[:, :2, :, :]

    fy = y * (1 - y)
    x = y + prompt * indicator * fy

    for outer_iter in arange(outer_itern):
        x0 = x
        if isinner == 1:
            for inner_iter in arange(inner_itern[outer_iter]):
                fx = x * (1 - x)
                x = x0 + prompt * indicator * fx
        else:
            x = x0

    if colorspace == 'RGB':
        enhance = tensor2img_numpy(x)
    elif colorspace == 'HSV':
        enhance = tensor2img_numpy_hsv(torch.cat((img_HS,x),1))


    enhance = np.ascontiguousarray(enhance)

    save_images(enhance, save_path, is_numpy=True)

    print(save_path + ' done!')

def main():
    for filename in os.listdir(path):
        prompt = 1
        img_path = path + '/' + filename
        save_path = outpath + '/' + filename

        img = np.array(Image.open(img_path))
        img = img / 255.0
        pixel_mean = np.mean(img)

        if pixel_mean > 0.5:
            prompt = -1
        
        pec(img_path, save_path, prompt)

if __name__ == '__main__':
    main()

File: Matlab/PEC.m

path = 'Input\under\';
outpath = 'Result\';
colorspace = 'HSV';  % 'RGB' or 'HSV' 
indicator = 0.55;  % c
outer_itern = 4;   % T
inner_itern = [1, 1, 1, 1];  % K
isinner = 1;

filenames = dir(fullfile(path, '*.JPG'));

for i = 1: length(filenames)
    prompt = 1;
    filename = filenames(i).name
    img_path = sprintf('%s%s', path, filenames(i).name);
    save_path = sprintf('%s%s', outpath, filenames(i).name);
    img = imread(img_path); 
    
    img = im2double(img); 

    pixel_mean = mean(img(:));
    if pixel_mean > 0.5
        prompt = -1;
    end
    % d = size(img)

    if strcmp(colorspace, 'RGB')  
        imgx = reshape(img, [1, size(img)]); 
        y = imgx; 

    elseif strcmp(colorspace, 'HSV')
        is_transition = true;
        imgx = rgb2hsv(img);
        imgx = permute(imgx, [3, 1, 2]); 
        imgx = reshape(imgx, [1, size(imgx)]); 
        y = squeeze(imgx(:, 3, :, :));  
        y = reshape(y, [1, size(y)]);

        y = reshape(y, [1, size(y)]); 
        img_HS = imgx(:, 1:2, :, :);  


    end


    fy = y .* (1 - y);
    x = y + prompt * indicator * fy;


    for outer_iter = 1: outer_itern
        x0 = x;
        if isinner == 1
            for inner_iter = 1: inner_itern(outer_iter)
                fx = x .* (1 - x);
                x = x0 + prompt * indicator * fx;
            end   
        else
            x = x0;
        end
    end


    if strcmp(colorspace, 'RGB')
        enhance = f_fuc(x, 1);
    elseif strcmp(colorspace, 'HSV')
        enhanced_img = cat(2, img_HS, x);
        enhance = f_fuc(enhanced_img, 2);
    end

    
    is_numpy = true;
    save_images(enhance, save_path, is_numpy);  
    
    fprintf('%s done!\n', save_path);  


end